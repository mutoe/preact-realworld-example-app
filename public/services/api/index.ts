import { apiService } from 'ts-api-toolkit';

apiService.changeBaseUrl('https://api.realworld.io/api');
apiService.changeAuthSchema('Token');

export function errorHandler(error: { data?: { [key: string]: string[] } }, fallbackMessage: string) {
	return error?.data?.errors ? error.data.errors : { unknown: [fallbackMessage] };
}

export { apiService };
