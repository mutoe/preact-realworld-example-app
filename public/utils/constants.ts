export const DEFAULT_AVATAR = 'https://static.realworld.io/images/smiley-cyrus.jpg';
