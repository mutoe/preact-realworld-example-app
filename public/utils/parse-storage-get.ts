export default function parseStorageGet(key: string) {
    try {
        const value = localStorage.getItem(key) || '';
        return JSON.parse(value);
    } catch (e) {
        return null;
    }
}
