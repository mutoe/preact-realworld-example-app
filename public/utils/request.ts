import parseStorageGet from './parse-storage-get';

interface FetchRequestOptions {
    prefix: string;
    headers: Record<string, any>;
    params: Record<string, any>;
    responseInterceptor: (response: Response) => void;
}

export default class FetchRequest {
    defaultOptions: FetchRequestOptions = {
        prefix: '',
        headers: {},
        params: {},
        responseInterceptor: (response) => response,
    };

    public options: FetchRequestOptions;

    constructor(options: Partial<FetchRequestOptions> = {}) {
        this.options = Object.assign({}, this.defaultOptions, options);
    }

    private generateFinalUrl = (url: string, options: Partial<FetchRequestOptions> = {}) => {
        const prefix = options.prefix || this.options.prefix || '';
        const params = options.params || {};

        let finalUrl = `${prefix}${url}`;
        if (Object.keys(params).length) {
            const queryString = Object.keys(params)
                .map((key) => `${key}=${params[key]}`)
                .join('&');
            finalUrl += `?${queryString}`;
        }

        return finalUrl;
    };

    private handleResponse = (response: Response) => {
        this.options.responseInterceptor(response);
        return response.json().then((json) => {
            if (response.status >= 200 && response.status < 300) {
                return json;
            }
            const error = new Error(response.statusText);
            Object.assign(error, json, {
                status: response.status,
                statusText: response.statusText,
            });
            throw error;
        });
    };

    get<T = any>(url: string, options: Partial<FetchRequestOptions> = {}): Promise<T> {
        options.headers = options.headers ?? {};
        const token = parseStorageGet('user')?.token;
        if (token) options.headers.Authorization = `Token ${token}`;

        const finalUrl = this.generateFinalUrl(url, options);
        return fetch(finalUrl, {
            method: 'GET',
            headers: this.options.headers,
        }).then(this.handleResponse);
    }

    post<T = any>(url: string, data: Record<string, any> = {}, options: Partial<FetchRequestOptions> = {}): Promise<T> {
        options.headers = options.headers ?? {};
        const token = parseStorageGet('user')?.token;
        if (token) options.headers.Authorization = `Token ${token}`;

        const finalUrl = this.generateFinalUrl(url, options);

        return fetch(finalUrl, {
            method: 'POST',
            body: JSON.stringify(data),
            headers: this.options.headers,
        }).then(this.handleResponse);
    }

    delete<T = any>(url: string, options: Partial<FetchRequestOptions> = {}): Promise<T> {
        options.headers = options.headers ?? {};
        const token = parseStorageGet('user')?.token;
        if (token) options.headers.Authorization = `Token ${token}`;

        const finalUrl = this.generateFinalUrl(url, options);

        return fetch(finalUrl, {
            method: 'DELETE',
            headers: this.options.headers,
        }).then(this.handleResponse);
    }

    put<T = any>(url: string, data: Record<string, any> = {}, options: Partial<FetchRequestOptions> = {}): Promise<T> {
        options.headers = options.headers ?? {};
        const token = parseStorageGet('user')?.token;
        if (token) options.headers.Authorization = `Token ${token}`;

        const finalUrl = this.generateFinalUrl(url, options);

        return fetch(finalUrl, {
            method: 'PUT',
            body: JSON.stringify(data),
            headers: this.options.headers,
        }).then(this.handleResponse);
    }

    patch<T = any>(
        url: string,
        data: Record<string, any> = {},
        options: Partial<FetchRequestOptions> = {},
    ): Promise<T> {
        options.headers = options.headers ?? {};
        const token = parseStorageGet('user')?.token;
        if (token) options.headers.Authorization = `Token ${token}`;

        const finalUrl = this.generateFinalUrl(url, options);

        return fetch(finalUrl, {
            method: 'PATCH',
            body: JSON.stringify(data),
            headers: this.options.headers,
        }).then(this.handleResponse);
    }
}
