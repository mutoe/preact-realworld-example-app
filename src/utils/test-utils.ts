import mockjs, { Random } from 'mockjs'
import { ReactWrapper } from 'enzyme'

export function generateProfile (): Profile {
  return mockjs.mock({
    username: Random.name(),
    bio: Random.sentence(),
    image: Random.image(),
    following: Random.boolean(),
  })
}

export function generateArticles(): Article
export function generateArticles(count: number): Article[]
export function generateArticles (count = 1): Article | Article[] {
  const title = Random.title()
  const slug = `${title.replace(/[ -]/g, '-')}-${Number(new Date()).toString(36)}`
  return mockjs.mock({
    [`articles|${count}`]: [
      {
        title,
        slug,
        body: Random.paragraph(),
        createdAt: new Date(Random.date()).toISOString(),
        updatedAt: new Date(Random.date()).toISOString(),
        'tagList|3': [() => Random.word()],
        description: Random.sentence(),
        author: () => generateProfile(),
        favorited: Random.boolean(),
        'favoritesCount|0-200': 200,
      },
    ],
  }).articles
}

export const setInputValue = (wrapper: ReactWrapper<any>, selector: string, value: string) => {
  wrapper.find(selector).getDOMNode<HTMLInputElement>().value = value
  wrapper.find(selector).simulate('input')
}

export const getInputValue = (wrapper: ReactWrapper<any>, selector: string): string => {
  return wrapper.find(selector).getDOMNode<HTMLInputElement>().value
}

export function generateComments(): ArticleComment
export function generateComments(count: number): ArticleComment[]
export function generateComments (count = 1): ArticleComment | ArticleComment[] {
  return mockjs.mock({
    [`comments|${count}`]: [
      {
        id: mockjs.Random.integer(0, 100),
        body: '@paragraph',
        createdAt: new Date(Random.date()).toISOString(),
        updatedAt: new Date(Random.date()).toISOString(),
        author: () => generateProfile(),
      },
    ],
  }).comments
}
