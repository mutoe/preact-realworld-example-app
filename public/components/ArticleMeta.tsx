import { useState } from 'preact/hooks';
import { useLocation } from 'preact-iso';

import { apiDeleteArticle, apiFavoriteArticle, apiUnfavoriteArticle } from '../services/api/article';
import { apiFollowProfile, apiUnfollowProfile } from '../services/api/profile';
import { useStore } from '../store';
import { dateFormatter } from '../utils/dateFormatter';
import { DEFAULT_AVATAR } from '../utils/constants';

interface ArticleMetaProps {
	article: Article;
}

export function ArticleMeta(props: ArticleMetaProps) {
	const [article, setArticle] = useState(props.article);
	const user = useStore(state => state.user);
	const location = useLocation();

	const onDelete = async () => {
		await apiDeleteArticle(article.slug);
		location.route('/');
	};

	const onFollow = async () => {
		const author = article.author.following
			? await apiUnfollowProfile(article.author.username)
			: await apiFollowProfile(article.author.username);
		setArticle({ ...article, author });
	};

	const onFavorite = async () =>
		setArticle(article.favorited ? await apiUnfavoriteArticle(article.slug) : await apiFavoriteArticle(article.slug));

	return (
		<div class="article-meta">
			<a href={`/@${article.author.username}`}>
				<img src={article.author.image || DEFAULT_AVATAR} alt="User's profile picture" />
			</a>
			<div class="info">
				<a href={`/@${article.author.username}`} class="author">
					{article.author.username}
				</a>
				<span class="date">{dateFormatter(article.createdAt)}</span>
			</div>
			{user?.username === article.author.username ? (
				<>
					<a class="btn btn-sm btn-outline-secondary" href={`/editor/${article.slug}`}>
						<i class="ion-edit" /> Edit Article
					</a>
					&nbsp;&nbsp;
					<button class="btn btn-sm btn-outline-danger" onClick={onDelete}>
						<i class="ion-trash-a" /> Delete Article
					</button>
				</>
			) : (
				<>
					<button
						class={`btn btn-sm ${article.author.following ? 'btn-secondary' : 'btn-outline-secondary'}`}
						onClick={onFollow}
					>
						<i class="ion-plus-round" /> {article.author.following ? 'Unfollow' : 'Follow'} {article.author.username}
					</button>
					&nbsp;&nbsp;
					<button
						class={`btn btn-sm ${article.favorited ? 'btn-primary' : 'btn-outline-primary'}`}
						onClick={onFavorite}
					>
						<i class="ion-heart" />
						&nbsp; Favorite Article <span class="counter">({article.favoritesCount})</span>
					</button>
				</>
			)}
		</div>
	);
}
