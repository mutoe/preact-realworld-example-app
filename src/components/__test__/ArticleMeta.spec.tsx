import render from 'preact-render-to-string'
import ArticleMeta from '../ArticleMeta'
import { h } from 'preact'
import { shallow } from 'enzyme'
import { generateArticles } from '../../utils/test-utils'
import { deleteFavoriteArticle, deleteFollowProfile, postFavoriteArticle, postFollowProfile } from '../../services'

jest.mock('../../services')

describe('# Article meta component', function () {
  const setArticle = jest.fn()

  it('should match snapshot', function () {
    const article = {
      title: 'title',
      slug: 'title-slug',
      body: 'body',
      createdAt: '1991-09-29T00:00:00.000Z',
      updatedAt: '2016-10-14T00:00:00.000Z',
      tagList: ['foo', 'bar', 'baz'],
      description: 'description',
      favorited: false,
      favoritesCount: 34,
      author: {
        username: 'Brenda Taylor',
        bio: 'bio.',
        image: 'http://dummyimage.com/234x60',
        following: false,
      },
    }
    const wrapper = render(<ArticleMeta article={article} setArticle={setArticle} />)

    expect(wrapper).toMatchSnapshot()
  })

  it('should follow user correctly', async function () {
    const article = generateArticles()
    article.author.following = false
    const wrapper = shallow(<ArticleMeta article={article} setArticle={setArticle} />)
    const followButton = wrapper.find('.ion-plus-round').parents('button')
    followButton.simulate('click')

    expect(postFollowProfile).toBeCalledTimes(1)
    expect(postFollowProfile).toBeCalledWith(article.author.username)
  })

  it('should unfollow user correctly', async function () {
    const article = generateArticles()
    article.author.following = true
    const wrapper = shallow(<ArticleMeta article={article} setArticle={setArticle} />)
    const followButton = wrapper.find('.ion-plus-round').parents('button')
    followButton.simulate('click')

    expect(deleteFollowProfile).toBeCalledTimes(1)
    expect(deleteFollowProfile).toBeCalledWith(article.author.username)
  })

  it('should favorite article correctly', function () {
    const article = generateArticles()
    article.favorited = false
    const wrapper = shallow(<ArticleMeta article={article} setArticle={setArticle} />)
    const favoriteButton = wrapper.find('.ion-heart').parents('button')
    favoriteButton.simulate('click')

    expect(postFavoriteArticle).toBeCalledTimes(1)
    expect(postFavoriteArticle).toBeCalledWith(article.slug)
  })

  it('should unfavored article correctly', function () {
    const article = generateArticles()
    article.favorited = true
    const wrapper = shallow(<ArticleMeta article={article} setArticle={setArticle} />)
    const favoriteButton = wrapper.find('.ion-heart').parents('button')
    favoriteButton.simulate('click')

    expect(deleteFavoriteArticle).toBeCalledTimes(1)
    expect(deleteFavoriteArticle).toBeCalledWith(article.slug)
  })
})
