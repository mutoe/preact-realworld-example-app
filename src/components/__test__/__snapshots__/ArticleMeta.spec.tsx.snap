// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`# Article meta component should match snapshot 1`] = `
<div class="article-meta">
  <a href="/@Brenda Taylor">
    <img src="http://dummyimage.com/234x60" />
  </a>
  <div class="info">
    <a href="/@Brenda Taylor" class="author">Brenda Taylor</a>
    <span class="date">September 29</span>
  </div>
  <button class="btn btn-sm btn-outline-secondary">
    <i class="ion-plus-round"></i>Follow Brenda Taylor</button>  
  <button class="btn btn-sm btn-outline-primary">
    <i class="ion-heart"></i>  Favorite Post
    <span class="counter">(34)</span>
  </button>
</div>
`;
